### PR TITLE
Add support for filtering metric label sets

### PR DESCRIPTION
**Pull request by jmacd**
This PR adds a `label.Set.Filter()` method used to reduce label set dimensionality in a metric Processor. The Filter is a `func(string) bool` that includes labels when true and excludes labels when false.

This was demonstrated in #1023, as being potentially useful in the Accumulator to avoid accumulation of synchronous instrument records by filtering at every metric event. Note that label filtering requires the fix in #1024 whether performed in the Accumulator or a Processor.

A companion PR #1048 adds a label set dimensionality-reducing Processor.

### CODE REVIEW

File: api/label/set.go [hunk 1]
```diff
@@ -116,8 +123,10 @@ func (l *Set) Value(k kv.Key) (kv.Value, bool) {
 	if l == nil {
 		return kv.Value{}, false
 	}
-	rValue := l.equivalent.reflect()
-	vlen := rValue.Len()
+	var vlen int
+	var rValue reflect.Value
+	vlen = rValue.Len()
+	rValue = l.equivalent.reflect()
```
Comment (Aneurysm9) [inline api/label/set.go:129]: Is the swapping of the initialization order here intentional?  Won't this initialize `vlen` to 0, resulting in `sort.Search` immediately returning 0 as it is searching on the range [0,0), which then causes the `if idx >= vlen` check to be the only case that can hit as `idx == vlen` will always be true?

Comment (jmacd) [reply api/label/set.go]: This was actually panicking! I've reversed those lines and added a test. Thanks! 😀

File: api/label/set.go [hunk 2]
```diff
@@ -50,6 +50,13 @@ type (
 		iface interface{}
 	}
 
+	// Filter supports removing certain keys from label sets.
+	// When the filter returns true, the label key will be kept in
+	// the filtered label set.  When the filter returns false, the
+	// label key is excluded from the filtered label set, and the
+	// value instead appears in the `removed` list of excluded keys.
+	Filter func(string) bool
```
[the remark this answers is not in the document]
Comment (jmacd) [reply api/label/set.go]: I can't see any significant downsides to the change you proposed. Done.
However, I didn't quite understand your meaning:

> That would mean that the included and removed key sets would not be unique though (not sure if that's an issue). 

(I don't think there's an issue.)

File: api/label/set.go [hunk 3]
```diff
@@ -50,6 +50,13 @@ type (
 		iface interface{}
 	}
 
+	// Filter supports removing certain keys from label sets.
+	// When the filter returns true, the label key will be kept in
+	// the filtered label set.  When the filter returns false, the
+	// label key is excluded from the filtered label set, and the
+	// value instead appears in the `removed` list of excluded keys.
```
Comment (MrAlias) [inline api/label/set.go:57]:
```suggestion
	// Filter supports removing certain labels from label sets.
	// When the filter returns true, the label will be kept in
	// the filtered label set.  When the filter returns false, the
	// label is excluded from the filtered label set, and the
	// label instead appears in the `removed` list of excluded labels.
```

File: api/label/set.go [hunk 4]
```diff
@@ -249,21 +287,22 @@ func NewSet(kvs ...kv.KeyValue) Set {
...
 //
 // - allocating a `Sortable` for use as a temporary in this method
 // - allocating a `Set` for storing the return value of this
 //   constructor.
 //
 // The result maintains a cache of encoded labels, by label.EncoderID.
 // This value should not be copied after its first use.
-func NewSetWithSortable(kvs []kv.KeyValue, tmp *Sortable) Set {
+//
+// The second `[]kv.KeyValue` return value is a list of keys that were
```
Comment (MrAlias) [inline api/label/set.go:300]:
```suggestion
// The second `[]kv.KeyValue` return value is a list of labels that were
```

File: api/label/set.go [hunk 5]
```diff
@@ -287,13 +326,55 @@ func NewSetWithSortable(kvs []kv.KeyValue, tmp *Sortable) Set {
...
 		equivalent: computeDistinct(kvs[position:]),
+	}, nil
+}
+
+// filterSet reorders `kvs` so that included keys are contiguous at
+// the end of the slice, while excluded keys precede the included keys.
+func filterSet(kvs []kv.KeyValue, filter Filter) (Set, []kv.KeyValue) {
+	var excluded []kv.KeyValue
+
+	// Move labels that do not match the optional regexp so
```
Comment (MrAlias) [inline api/label/set.go:345]:
```suggestion
	// Move labels that do not match the filter so
```

**Pull request merged**